// aws-cdk/test/tree.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable import/order */
const path = require("path");
const tree_1 = require("../lib/tree");
describe('some', () => {
    const tree = {
        id: 'App',
        path: '',
        children: {
            Tree: {
                id: 'Tree',
                path: 'Tree',
                constructInfo: {
                    fqn: 'aws-cdk-lib.Construct',
                    version: '1.162.0',
                },
            },
            stack: {
                id: 'stack',
                path: 'stack',
                children: {
                    bucket: {
                        id: 'bucket',
                        path: 'stack/bucket',
                        children: {
                            Resource: {
                                id: 'Resource',
                                path: 'stack/bucket/Resource',
                                attributes: {
                                    'aws:cdk:cloudformation:type': 'AWS::S3::Bucket',
                                    'aws:cdk:cloudformation:props': {},
                                },
                                constructInfo: {
                                    fqn: '@aws-cdk/aws-s3.CfnBucket',
                                    version: '1.162.0',
                                },
                            },
                        },
                        constructInfo: {
                            fqn: '@aws-cdk/aws-s3.Bucket',
                            version: '1.162.0',
                        },
                    },
                    CDKMetadata: {
                        id: 'CDKMetadata',
                        path: 'stack/CDKMetadata',
                        children: {
                            Default: {
                                id: 'Default',
                                path: 'stack/CDKMetadata/Default',
                                constructInfo: {
                                    fqn: 'aws-cdk-lib.CfnResource',
                                    version: '1.162.0',
                                },
                            },
                            Condition: {
                                id: 'Condition',
                                path: 'stack/CDKMetadata/Condition',
                                constructInfo: {
                                    fqn: 'aws-cdk-lib.CfnCondition',
                                    version: '1.162.0',
                                },
                            },
                        },
                        constructInfo: {
                            fqn: 'aws-cdk-lib.Construct',
                            version: '1.162.0',
                        },
                    },
                },
                constructInfo: {
                    fqn: 'aws-cdk-lib.Stack',
                    version: '1.162.0',
                },
            },
        },
        constructInfo: {
            fqn: 'aws-cdk-lib.App',
            version: '1.162.0',
        },
    };
    test('tree matches predicate', () => {
        expect((0, tree_1.some)(tree, node => node.constructInfo?.fqn === '@aws-cdk/aws-s3.Bucket')).toBe(true);
    });
    test('tree does not match predicate', () => {
        expect((0, tree_1.some)(tree, node => node.constructInfo?.fqn === '@aws-cdk/aws-lambda.Function')).toBe(false);
    });
    test('childless tree', () => {
        const childless = {
            id: 'App',
            path: '',
            constructInfo: {
                fqn: 'aws-cdk-lib.App',
                version: '1.162.0',
            },
        };
        expect((0, tree_1.some)(childless, node => node.path.length > 0)).toBe(false);
    });
});
describe('loadTreeFromDir', () => {
    test('can find tree', () => {
        const tree = (0, tree_1.loadTreeFromDir)(path.join(__dirname, 'cloud-assembly-trees', 'built-with-1_144_0'));
        expect(tree.id).toEqual('App');
    });
    test('cannot find tree', () => {
        const tree = (0, tree_1.loadTreeFromDir)(path.join(__dirname, 'cloud-assembly-trees', 'foo'));
        expect(tree).toEqual({});
    });
});
